server/pkcs11.go

```diff
@@ -0,0 +1,182 @@
+package server
+
+import (
+	"crypto"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cbroglie/crypto11"
+)
+
+// A PKCS11URI represents a PKCS#11 object stored on a PKCS#11 token and
+// information for accessing it. The general form represented is:
+//
+//	pkcs11:path-attr[?query-attr]
+//
+// Path and query attributes are delimited by ';' and '&', respectively.
+// For more information read: https://tools.ietf.org/html/rfc7512#section-2.3
+type PKCS11URI struct {
+	// path attributes:
+	Token  string //        token <- CK_TOKEN_INFO
+	Manuf  string // manufacturer <- CK_TOKEN_INFO
+	Serial string //       serial <- CK_TOKEN_INTO
+	Model  string //        model <- CK_TOKEN_INFO
+
+	LibManuf string // library-manufacturer <- CK_INFO
+	LibDesc  string //  library-description <- CK_INFO
+	LibVer   string //      library-version <- CK_INFO
+
+	Object []byte // object <- CKA_LABEL
+	Type   string //   type <- CKA_CLASS
+	ID     []byte //     id <- CKA_ID
+
+	SlotManuf string // slot-manufacturer <- CK_SLOT_INFO
+	SlotDesc  string //  slot-description <- CK_SLOT_INFO
+	SlotID    uint   //           slot-id <- CK_SLOT_ID
+
+	// query attributes:
+	PinSource string // pin-source
+	PinValue  string //  pin-value
+
+	ModuleName string // module-name
+	ModulePath string // module-path
+
+	// Vendor specific query attributes:
+	MaxSessions int // max-sessions
+}
+
+// PKCS11Parser decodes a PKCS#11 URI as defined in RFC 7512 into an PKCS11URI object.
+// An error is returned if the input string does not appear to follow the rules
+// or if there are unrecognized path or query attributes.
+func PKCS11Parser(uri string) (*PKCS11URI, error) {
```

Comment:
The function name should be declarative. I would prefer 'ParseURI' and we change PKCS11URI to just URI.

Reply:
I disagree, unless you want me to remove this code and make a `rfc7512` package, then `rfc7512.ParseURI` and `rfc7512.URI` would be appropriate.